handlers/definitionsForRoles.js

```diff
@@ -27,86 +27,87 @@ module.exports = {
-               ).length > 0;
-
-            var ids = [];
-            // {array}
-            // all the ABDefinition.id that need to be exported.
-
-            var hashKey = roleIDs.join(",");
-            if (!hashIDs[hashKey] || hashIDs[hashKey].length == 0) {
-               req.log("building ID hash");
-               req.performance.mark("buildIDHash");
-               var applications = AB.applications(
-                  (a) => isSystemUser || a.isAccessibleForRoles(roles)
-               );
-
-               req.log(
-                  `definition_manager.definitionsForRoles: found ${applications.length} applications to export`
-               );
-
-               // This takes a long time!
-               // Cache this?
-               let aIDs = [];
-               applications.forEach((a) => {
-                  a.exportIDs(aIDs);
-               });
-
-               // NOTE: we also need to make sure all the System Objects in the definitions.
-               let systemObjects = AB.objects((o) => o.isSystemObject);
-               for (var i = 0; i < systemObjects.length; i++) {
-                  systemObjects[i].exportIDs(aIDs);
-               }
-
-               hashIDs[hashKey] = aIDs;
-
-               AB.cache("defs-for-role", hashIDs);
-               req.performance.measure("buildIDHash");
-            }
-
-            ids = hashIDs[hashKey];
-            req.log(
-               `definition_manager.definitionsForRoles: found ${ids.length} ids to export.`
-            );
-            var definitions = [];
-            ids.forEach((id) => {
-               let def = AB.definitionByID(id, true);
-               if (def) {
-                  definitions.push(def);
-               }
-            });
-
-            cb(null, definitions);
-         })
-         .catch((err) => {
-            // we clear the cache just in case our data was incorrect.
-            // AB.cacheClear(ServiceKey);
-            req.notify.developer(err, {
-               context:
-                  "Service:definition_manager.definitionsForRoles: Error initializing ABFactory",
-            });
-            cb(err);
+      try {
+         const AB = await ABBootstrap.init(req);
+         const stringifiedDefs = AB.cache("cached-defs") || {};
+         const hashKey = roleIDs.join(",");
+
+         if (stringifiedDefs[hashKey] != null) {
+            cb(null, stringifiedDefs[hashKey]);
+
+            return;
+         }
+
+         req.log("building ID hash");
+         req.performance.mark("buildIDHash");
+
+         const isSystemUser =
+            roleIDs.filter((roleId) => AB.defaultSystemRoles().includes(roleId))
+               .length > 0;
+         const applications = AB.applications(
+            (a) => isSystemUser || a.isAccessibleForRoles(roles)
+         );
+
+         req.log(
+            `definition_manager.definitionsForRoles: found ${applications.length} applications to export`
+         );
+
+         // This takes a long time!
```

Comment:
We've just done the important caching in line `49` so we can remove these comments.

Reply:
https://github.com/digi-serve/ab_service_definition_manager/pull/64/commits/972825f143d3ab4cd12d151c9e268bb93e1748a1